### PR TITLE
[soc] Add FPGA Sim Files

### PR DESCRIPTION
**Pull request by cyyself**

### CODE REVIEW

**Comment by Avimitin** [comment]
I've fixed the nix coding style, please update your local branch.

File: nix/t1/soc-elaborate.nix [hunk 1]
```diff
@@ -7,6 +7,7 @@
 , elaborate-config
 , soc-elaborator
 , config-name
+, enableFpga ? false
```
Comment (sequencer) [inline nix/t1/soc-elaborate.nix:10]: no default value plz

File: subsystememu/csrc/fpga/FPGASim.cpp [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        printf("soc and cpu reseted!\n");
+    }
+
+    void dump_debug() {
+        printf("cpu reset=%x\n", user_bar->cpu_reset);
+        printf("last mem read=%x\n", user_bar->lastMemRead);
+        printf("last mem write=%x\n", user_bar->lastMemWrite);
+        printf("last mmio read=%x\n", user_bar->lastMMIORead);
+        printf("last mmio write=%x\n", user_bar->lastMMIOWrite);
+	    printf("has mmio error=%x\n", user_bar->hasMMIOError);
```
Comment (sequencer) [inline subsystememu/csrc/fpga/FPGASim.cpp:110]: nit: space

File: subsystememu/src/Main.scala [hunk 3]
```diff
@@ -41,6 +43,7 @@ object Main {
         case a => Some(a)
       }
     os.write(dir_ / s"$topName.anno.json", firrtl.annotations.JsonProtocol.serialize(annos))
+    freechips.rocketchip.util.ElaborationArtefacts.files.foreach { case (ext, contents) => os.write.over(dir_ / s"${p.toString}.${ext}", contents()) }
```
Comment (sequencer) [inline subsystememu/src/Main.scala:46]: do we need this?

File: subsystememu/src/Subsystem.scala [hunk 4]
```diff
@@ -72,7 +78,7 @@ class VerdesConfig
       .orElse(new WithCacheBlockBytes(16))
       // SoC
       .orElse(new WithoutTLMonitors)
-      .orElse(new WithDebugSBA)
+      .orElse(new WithNExtTopInterrupts(1))
```
Comment (sequencer) [inline subsystememu/src/Subsystem.scala:81]: do we need it?

File: subsystememu/src/fpga/MidShell.scala [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+  val lastMMIOWrite = withReset(cpuResetOut) { RegInit(~0.U(32.W)) }
+  val hasMMIOError = withReset(cpuResetOut) { RegInit(false.B) }
+  when(io.mem.ar.fire) {
+    lastMemRead := io.mem.ar.bits.addr
+  }
+  when(io.mem.aw.fire) {
+    lastMemWrite := io.mem.aw.bits.addr
+  }
+
+  def doRead(addr: UInt): (UInt, Bool) = {
```
Comment (sequencer) [inline subsystememu/src/fpga/MidShell.scala:89]: dirty... but ok...

File: subsystememu/src/fpga/VCU118Harness.scala [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+  val s1_available = withClockAndReset(ddr.io.axi_clock, ddr.io.axi_reset) {
+    RegNext(!axiintc.io.s1_axi_reset.asBool)
+  }
+  val socAvailable = withClockAndReset(midShellClock, midShellReset) {
+    RegNext(RegNext(s0_available, init=false.B) && RegNext(s1_available, init=false.B), false.B)
+  }
+  xdma.io.irq_req := midShell.io.irq
+  midShell.io.socAvailable := socAvailable
+
+  ElaborationArtefacts.add( desiredName + ".xdc",
```
Comment (sequencer) [inline subsystememu/src/fpga/VCU118Harness.scala:96]: I see...

**Review by sequencer — COMMENTED** [review]
a little nitpick

File: subsystememu/src/Main.scala (hunk 3 above)
Comment (cyyself) [reply subsystememu/src/Main.scala:46]: Yes. We need some `.tcl` scripts generated by using ElaborationArtefacts.

File: subsystememu/src/Subsystem.scala (hunk 4 above)
Comment (cyyself) [reply subsystememu/src/Subsystem.scala:81]: This is useful if we need UART input without polling for debug purposes.

File: subsystememu/csrc/fpga/FPGASim.cpp [hunk 7]
```diff
@@ -0,0 +1,172 @@
+#define XDMA_CHARDEV_TX "/dev/xdma0_h2c_0"
+#define XDMA_CHARDEV_RX "/dev/xdma0_c2h_0"
+#define XDMA_USER "/dev/xdma0_user"
```
Comment (sequencer) [inline subsystememu/csrc/fpga/FPGASim.cpp:3]: This should be extracted out.

Comment (sequencer) [reply subsystememu/csrc/fpga/FPGASim.cpp:3]: But OK for now.

**Review by sequencer — APPROVED** [review]

**Pull request merged**